File: Algorithms/Python/0113_Path_Sum_II/Path_Sum_II_Solution_2.py
# Space: O(n)
# Time: O(n)

# BFS approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def pathSum(self, root, s):
        if root is None: return []
        res, path = [], []

        queue = [[root, path]]

        while queue:
            cur_root, cur_path = queue.pop(0)
            if cur_root.left is None and cur_root.right is None and s == sum(cur_path) + cur_root.val:
                res.append(cur_path + [cur_root.val])

            if cur_root.left:
                queue.append([cur_root.left, cur_path + [cur_root.val]])
            if cur_root.right:
                queue.append([cur_root.right, cur_path + [cur_root.val]])

        return res

File: Algorithms/0113_Path_Sum_II/Java/PathSumIISolution2.java

// Space: O(n)
// Time: O(n)

class Solution {
    List<List<Integer>> res = null;

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        if (root == null) return new ArrayList<>();
        res = new ArrayList<>();
        dfs(root, new ArrayList<>(), targetSum);
        return res;
    }

    private void dfs(TreeNode root, List<Integer> temp, int target) {
        temp.add(root.val);
        if (root.left == null && root.right == null) {
            if (temp.stream().mapToInt(Integer::intValue).sum() == target) {
                res.add(new ArrayList<>(temp));
            }
            temp.remove(temp.size() - 1);
            return;
        }
        if (root.left != null) dfs(root.left, temp, target);
        if (root.right != null) dfs(root.right, temp, target);
        temp.remove(temp.size() - 1);
    }
}




